materialize-slack/main.go

```diff
@@ -0,0 +1,302 @@
+		bindings: bindings,
+	}
+
+	return transactor, &pm.Response_Opened{}, nil
+}
+
+type transactor struct {
+	api      *SlackAPI
+	bindings []*binding
+}
+
+type binding struct {
+	spec       *pf.MaterializationSpec_Binding
+	resource   resource
+	collection string
+}
+
+func buildDocument(b *binding, keys, values tuple.Tuple) map[string]interface{} {
+	var document = make(map[string]interface{})
+
+	// Add the keys to the document.
+	for i, value := range keys {
+		var propName = b.spec.FieldSelection.Keys[i]
+		document[propName] = value
+	}
+
+	// Add the non-keys to the document.
+	for i, value := range values {
+		var propName = b.spec.FieldSelection.Values[i]
+
+		if raw, ok := value.([]byte); ok {
+			document[propName] = json.RawMessage(raw)
+		} else {
+			document[propName] = value
+		}
+	}
+	return document
+}
+
+func (d *transactor) Load(it *pm.LoadIterator, loaded func(int, json.RawMessage) error) error {
+	for it.Next() {
+		panic("Load should never be called for materialize-slack")
+	}
+	return nil
+}
+
+func (t *transactor) Store(it *pm.StoreIterator) (pm.StartCommitFunc, error) {
+
+	for it.Next() {
+		var b = t.bindings[it.Binding]
+		var parsed = buildDocument(b, it.Key, it.Values)
+		var tsStr, tsOk = parsed["ts"].(string)
+		var text, textOk = parsed["text"].(string)
+		var blocks, blocksOk = parsed["blocks"].(json.RawMessage)
+
+		if !tsOk {
+			return nil, fmt.Errorf("missing timestamp")
+		}
+
+		if !textOk {
+			return nil, fmt.Errorf("missing text")
+		}
+
+		// Parse the timestamp as a time.Time
+		ts, err := time.Parse(time.RFC3339Nano, tsStr)
+		if err != nil {
+			return nil, fmt.Errorf("invalid timestamp %q", tsStr)
+		}
+
+		var blocksParsed slack.Blocks
+		if blocksOk {
+			// Is this jank? This should really be taken care of by the
+			// serialization logic in slack.Block, but it can't be _that_
+			// bad to do it here... right? see below:
+			// https://api.slack.com/reference/surfaces/formatting#escaping
+			var escaped_blocks = string(blocks)
+			escaped_blocks = strings.ReplaceAll(escaped_blocks, "&", "&amp;")
+			escaped_blocks = strings.ReplaceAll(escaped_blocks, "<", "&lt;")
+			escaped_blocks = strings.ReplaceAll(escaped_blocks, ">", "&gt;")
+
+			err = json.Unmarshal([]byte(escaped_blocks), &blocksParsed)
+
+			if err != nil {
+				return nil, fmt.Errorf("invalid blocks value %q, %q: %w", reflect.TypeOf(blocks), string(blocks), err)
+			}
+		} else {
+			log.Warn(fmt.Sprintf("Blocks apparently NOT okay: %+v", reflect.TypeOf(parsed["blocks"])))
+		}
+
+		// Accept messages from at most 10 minutes in the past
+		if time.Since(ts).Minutes() < 10 {
+			if err := t.api.PostMessage(b.resource.Channel, text, blocksParsed.BlockSet, b.resource.SenderConfig); err != nil {
+				serializedBlocks, marshal_err := json.Marshal(blocksParsed)
+				if marshal_err == nil {
+					log.Warn(fmt.Sprintf("Parse Blocks: %+v", string(serializedBlocks)))
+				}
+				log.Warn(fmt.Errorf("error sending message: %w", err))
```

Comment:
This may be something where we'd want to retry with rate limiting, but it doesn't look like that is implemented here, although I see the partial comment below and the sleep.

Reply:
Yeah, retries would be nice. Do we have any common way of implementing that with e.g exponential backoff and all that?

Reply:
You could probably leverage something like this, from the google sheets materialization: https://github.com/estuary/connectors/blob/0f94d077199f097af12f588dcc491a65e1e922be/materialize-google-sheets/util.go#L305-L327